### PR TITLE
Challenge 2 - Pythagorean Triplets Finder

### PR DESCRIPTION
**Pull request by malakbattat**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [X] The file name describes the function's behavior
- [X] There is a module docstring in the function file
- [X] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [X] There is a module docstring in the tests file

### Unit Tests

- [X] The test class has a helpful name in PascalCase
- [X] The test class has a docstring
- Every unit test has
  - [X] A helpful name
  - [X] A clear docstring
  - [X] Only one assertion
  - [X] There is no logic in the unit test
- [X] All tests pass
- [X] There are tests for defensive assertions
- [X] There are tests for boundary cases

### Function Docstring

- [X] The function's behavior is described
- The function's arguments are described:
  - [X] Type
  - [X] Purpose
  - [X] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [X] Type
  - [X] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [X] Each assumption about an argument is checked with an assertion
  - [X] Each assertion checks for _only one_ assumption about the argument
- [X] Include 3 or more (passing!) doctests

### The Function

- [X] The function's name describes it's behavior
- [X] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [X] The function has correct type annotations
- [X] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [X] Variable names help to understand the strategy
- [X] Any comments are clear and describe the strategy
- [X] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [X] The function's strategy _is not_ described in any docstrings or tests
- [X] Comments explain the _strategy_, **not** the _implementation_
- [X] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [X] The code passes the formatting checks
- [X] The code passes all Ruff linting checks
- [X] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [X] Variables are named with snake_case
- [X] Variable names are clear and helpful
- [X] The code follows the strategy as simply as possible
- [X] The implementation is as simple as possible given the strategy
- [X] There are no commented lines of code
- [X] There are no `print` statements anywhere
- [X] The code includes defensive assertions
- [X] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by Aseel-AbuKmail — COMMENTED** [review]
Hi Malak, it's a pleasure to review your code. Your implementation is clear and well-structured! However, as a reviewer, I have a few suggestions to further improve your code.

**Comment by malakbattat** [comment]
> Hi Malak, it's a pleasure to review your code. Your implementation is clear and well-structured! However, as a reviewer, I have a few suggestions to further improve your code.

Thank you so much Aseel! 
I'd love to hear your suggestions

**Comment by malakbattat** [comment]
Thank you so much for your review and suggestions Aseel!
I'm working on your suggestions at the moment

File: solutions/tests/test_find_pythagorean_triplets.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import os
+import sys
+import unittest
+
+# the tests are passing even with unable to import warning/error
+from ..find_pythagorean_triplets import find_primitive_pythagorean_triplets
+
+# Add the parent directory to the system path
+sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
```
[the remark this answers is not in the document]
Comment (malakbattat) [reply solutions/tests/test_find_pythagorean_triplets.py:22]: Thank you for the suggestion but unfortunately the imports don't work correctly without this line

File: solutions/tests/test_find_pythagorean_triplets.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    def test_multiple_triplets(self):
+        """Test with a range that includes multiple triplets"""
+        self.assertEqual(
+            find_primitive_pythagorean_triplets(20),
+            [(3, 4, 5), (5, 12, 13), (8, 15, 17)],
+        )
+
+    def test_no_triplets(self):
+        """Test with a range that includes no triplets"""
+        self.assertEqual(find_primitive_pythagorean_triplets(2), [])
```
[the remark this answers is not in the document]
Comment (malakbattat) [reply solutions/tests/test_find_pythagorean_triplets.py:40]: I couldn't find a way to take your suggestion and still optimize my code based on it, it was taking too long in running the test

[the remark this answers is not in the document]
Comment (Aseel-AbuKmail) [reply solutions/tests/test_find_pythagorean_triplets.py:40]: I understand that my suggestion might not have worked as intended for your use case. Optimizing for performance and memory usage can indeed be challenging. If you could share more details about the specific bottlenecks or constraints you're facing, I'd be happy to refine my approach or explore alternative solutions with you.
